File: day10/part2.java
import java.io.IOException;
import java.nio.file.*;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;
//https://adventofcode.com/2020/day/10
public class part2 {
    public static void main(String... args) throws IOException {
        List<Integer >numbers = Files.lines(Path.of("input.txt"))
                                     .map(Integer::valueOf)
                                     .sorted()
                                     .collect(Collectors.toList());

        printNumberOfWaysToConnectAdapters(numbers);
    }

    private static void printNumberOfWaysToConnectAdapters(List<Integer> numbers) {
        int max = numbers.stream()
                         .mapToInt(i -> i)
                         .max()
                         .orElseThrow(NoSuchElementException::new);

        numbers.add(numbers.size(),max+3);

        long ans = permutations(numbers);
        System.out.println(ans);
    }

    private static long permutations(List<Integer>list){
        long ans = 1;
        int ones = 1;

        for(int i = 1; i < list.size(); i++){
            int diff = list.get(i) - list.get(i-1);

            ones += diff == 1 ? 1 : 0;
            ans *= diff >= 3 ? sequence(ones) : 1;
            ones = diff >= 3 ? 0 : ones;
        }    
        return ans;
    }
    private static long sequence(int n){
        if(n >= 3)
            return sequence(n-1) + sequence(n-2) + sequence(n-3);
        if(n == 2)
            return 2;
        else
            return 1;
    }
}

File: day09/part2.cpp
#include<iostream>
#include<fstream>
#include<algorithm>
#include<vector>
//https://adventofcode.com/2020/day/9
using namespace std;
 
int main(){
    ifstream input;
    long long max, min, sum, numbers, ans;
    input.open("input.txt");
    vector<long long> v1(1000);

    for(long long &x : v1) input >> x;

    for(int i = 0; i < v1.size(); i++){
        sum = v1[i];
        min = max;
        max = 0;
        for(int j = i+1; j < v1.size(); j++){
            sum += v1[j];
            max = max < v1[j] ? v1[j] : max;
            min = min > v1[j] ? v1[j] : min;
            if(sum == 85848519) 
                ans = min + max;
        }
    } 
    cout << ans << endl; 
}